optuna/_gp/optim_mixed.py

```diff
@@ -0,0 +1,320 @@
+from __future__ import annotations
+
+import math
+from typing import TYPE_CHECKING
+
+import numpy as np
+
+from optuna._gp.acqf import AcquisitionFunctionParams
+from optuna._gp.acqf import eval_acqf_no_grad
+from optuna._gp.acqf import eval_acqf_with_grad
+from optuna._gp.search_space import normalize_one_param
+from optuna._gp.search_space import sample_normalized_params
+from optuna._gp.search_space import ScaleType
+from optuna.logging import get_logger
+
+
+if TYPE_CHECKING:
+    import scipy.optimize as so
+else:
+    from optuna import _LazyImport
+
+    so = _LazyImport("scipy.optimize")
+
+_logger = get_logger(__name__)
+
+
+def _gradient_ascent(
+    acqf_params: AcquisitionFunctionParams,
+    initial_params: np.ndarray,
+    initial_fval: float,
+    continuous_indices: np.ndarray,
+    lengthscale: np.ndarray,
+    tol: float,
+) -> tuple[np.ndarray, float, bool]:
+    """
+    This function optimizes the acquisition function using preconditioning.
+    Preconditioning equalizes the variances caused by each parameter and
+    speeds up the convergence.
+
+    In Optuna, acquisition functions use Matern 5/2 kernel, which is a function of `x / l`
+    where `x` is `normalized_params` and `l` is the corresponding lengthscales.
+    Then acquisition functions are a function of `x / l`, i.e. `f(x / l)`.
+    As `l` has different values for each param, it makes the function ill-conditioned.
+    By transforming `x / l` to `zl / l = z`, the function becomes `f(z)` and has
+    equal variances w.r.t. `z`.
+    So optimization w.r.t. `z` instead of `x` is the preconditioning here and
+    speeds up the convergence.
+    As the domain of `x` is [0, 1], that of `z` becomes [0, 1/l].
+    """
+    if len(continuous_indices) == 0:
+        return (initial_params, initial_fval, False)
+    normalized_params = initial_params.copy()
+
+    def negative_acqf_with_grad(scaled_x: np.ndarray) -> tuple[float, np.ndarray]:
+        # Scale back to the original domain, i.e. [0, 1], from [0, 1/s].
+        normalized_params[continuous_indices] = scaled_x * lengthscale
+        (fval, grad) = eval_acqf_with_grad(acqf_params, normalized_params)
+        # Flip sign because scipy minimizes functions.
+        # Let the scaled acqf be g(x) and the acqf be f(sx), then dg/dx = df/dx * s.
+        return (-fval, -grad[continuous_indices] * lengthscale)
+
+    scaled_cont_x_opt, neg_fval_opt, info = so.fmin_l_bfgs_b(
+        func=negative_acqf_with_grad,
+        x0=normalized_params[continuous_indices] / lengthscale,
+        bounds=[(0, 1 / s) for s in lengthscale],
+        pgtol=math.sqrt(tol),
+        maxiter=200,
+    )
+
+    if -neg_fval_opt < initial_fval or info["nit"] == 0:
+        # `nit` is the number of iterations.
+        return (initial_params, initial_fval, False)
+    else:
+        normalized_params[continuous_indices] = scaled_cont_x_opt * lengthscale
+        return (normalized_params, -neg_fval_opt, True)
```

Comment:
Shall we remove zero improvement? (The last one included zero improvement in updated.)
And just to reduce the same mistake in the equality stuff, I just follow this format for both exhaustive search and line search as well.

```suggestion
    if -neg_fval_opt > initial_fval and info["nit"] > 0:  # Improved.
        # `nit` is the number of iterations.
        normalized_params[continuous_indices] = scaled_cont_x_opt * lengthscale
        return (normalized_params, -neg_fval_opt, True)

    return (initial_params, initial_fval, False)  # No improvement.
```